pepper-apis/src/main/java/org/broadinstitute/ddp/util/RouteUtil.java

```diff
@@ -149,4 +159,18 @@ public static ActivityInstanceDto findAccessibleInstanceOrHalt(Response response
 
         return instanceDto;
     }
+
+    /**
+     * Figures out a preferred user language taking into account the language weights
+     * from Accept-Language header, information from the user profile and languages
+     * supported by the study.
+     */
+    public static Locale resolvePreferredUserLanguage(
+            Handle handle, String acceptLanguageHeader, Locale preferredLocale, String studyGuid
+    ) {
+        List<LanguageRange> acceptLanguages = StringUtils.isNotEmpty(acceptLanguageHeader)
+                ? LanguageRange.parse(acceptLanguageHeader) : new ArrayList<>();
+        Set<Locale> localesSupportedByStudy = handle.attach(StudyDao.class).getSupportedLocalesByGuid(studyGuid);
```

Comment:
Hmmm, is there a way to have `StudyDao` return a set of `LanguageCode` objects which has _both_ `iso_language_code` and `language_code_id`? So that we don't have to do yet another database query in the routes to `getLanguageCodeId`? Maybe something like
```java
class LangCode {
    private long langId;
    private String isoCode;
    ...
    public asLocale() {
      return Locale.forLanguageTag(isoCode);
    }
}
```

Reply:
I'd like us to do this whenever possible--those repeated calls for uk -> pk eventually will add up.

Reply:
That's what I was also thinking about when writing this code. Will change.

Reply:
We do it now.